### PR TITLE
Control shared/static linking via BUILD_SHARED_LIBS

### PR DESCRIPTION
**Pull request by esteve**
This PR delegates the decision to build a shared or a static library to CMake's BUILD_SHARED_LIBS

Update:

Connects to https://github.com/ros2/ros2/issues/306

### CODE REVIEW

File: rosidl_generator_c/CMakeLists.txt [hunk 1]
```diff
@@ -8,10 +8,11 @@ endif()
 
 find_package(ament_cmake REQUIRED)
 find_package(ament_cmake_python REQUIRED)
+find_package(ament_cmake_ros REQUIRED)
```
Comment (dirk-thomas) [inline rosidl_generator_c/CMakeLists.txt]: With `ament_cmake_ros` being added you can remove `ament_cmake` (which is a transitive dependency) to keep the CMake file simpler.

File: rosidl_generator_c/package.xml [hunk 2]
```diff
@@ -9,6 +9,7 @@
 
   <buildtool_depend>ament_cmake</buildtool_depend>
 
+  <build_depend>ament_cmake_ros</build_depend>
```
Comment (dirk-thomas) [inline rosidl_generator_c/package.xml]: This needs to be a `buildtool_depend` dependency.

With the above comment the `buildtool_depend` entry for `ament_cmake` can be removed.

File: rosidl_generator_c/CMakeLists.txt (hunk 1 above)
Comment (esteve) [inline rosidl_generator_c/CMakeLists.txt]: Done.

File: rosidl_generator_c/package.xml (hunk 2 above)
Comment (esteve) [inline rosidl_generator_c/package.xml]: Done.

**Comment by esteve** [comment]
I've also updated all the other PRs to use `ament_cmake_ros`

**Comment by dirk-thomas** [comment]
The typesupport libraries are expected to be shared when there is more then one available since they are being loaded using `dlopen`. When building static libraries but more than one typesupport is being used by `rosidl_typesupport_c|cpp` I would assume it fails at least at runtime. Have you tried that use case? I think it would be good if this case is being catched at build time to let the user know they are trying an impossible combination.

**Comment by esteve** [comment]
@dirk-thomas sorry, I forgot about this. Unfortunately I've only tested this with `rosidl_typesupport_introspection_c` and `rosidl_typesupport_introspection_cpp` through Fast-RTPS.

The changes need to happen somewhere around here, is that correct?

https://github.com/ros2/rosidl_typesupport/blob/master/rosidl_typesupport_c/cmake/rosidl_typesupport_c_generate_interfaces.cmake#L115-L126
https://github.com/ros2/rosidl_typesupport/blob/master/rosidl_typesupport_cpp/cmake/rosidl_typesupport_cpp_generate_interfaces.cmake#L101-L112

so that if there's more than one typesupport and the `BUILD_SHARED_LIBS` option is set to `OFF`, CMake would bail out, right? If so, let me know and I'll submit a PR to abort or warn the user that they can't use static libraries with more than one typesupport.

Thanks.

**Comment by dirk-thomas** [comment]
I am not sure if there is any use in building all libraries static when requested by still build the typesupport libraries shared if more than one is available?

If that is the case it would be best to choose the library type in each typesupport package to either use `BUILD_SHARED_LIBS` implicitly or if there is multiple typesupport explicitly build the typesupport libraries as `SHARED.

If not, the locations you referenced look like the right place to implement the bailing out.

**Comment by wjwwood** [comment]
> I am not sure if there is any use in building all libraries static when requested by still build the typesupport libraries shared if more than one is available?

Can you rephrase this, I don't exactly understand what you're saying. Maybe that it doesn't make sense to build static libraries if one than one type support is requested/possible?

**Comment by wjwwood** [comment]
edited tha last post, to fix my own typo

**Comment by dirk-thomas** [comment]
> Can you rephrase this, I don't exactly understand what you're saying. Maybe that it doesn't make sense to build static libraries if more than one type support is requested/possible?

The request to build static libraries is problematic if multiple typesupports are available. Multiple typesupports can only be supported with shared libraries. The question is what should happen in that case:

* build typesupport shared despite the request for static libraries
* build static typesupport which implies selecting a single one (which one? how to notify the user?)
* fail the build and ask the user to either switch to shared libraries or reduce the set of typesupports to only one

**Comment by wjwwood** [comment]
When you say multiple type support do you mean `rosidl_typesupport_introspection_c` and `rosidl_typesupport_introspection_cpp` would cause this issue or is it only an issue when "instrospection" and "connext static" (for example) are both requested? Put another way, is this an issue when you have C and C++ support or only when you request two rmw implementations that use two different type support schemes?

> - build typesupport shared despite the request for static libraries

Is that an acceptable solution @esteve? It seems to me like it wouldn't be in this case.

**Comment by dirk-thomas** [comment]
> When you say multiple type support do you mean rosidl_typesupport_introspection_c and rosidl_typesupport_introspection_cpp would cause this issue or is it only an issue when "instrospection" and "connext static" (for example) are both requested? Put another way, is this an issue when you have C and C++ support or only when you request two rmw implementations that use two different type support schemes?

The typesupport dispatch is happening in `rosidl_typesupport_c` as well as in `rosidl_typesupport_cpp`.  Each of them can only use a single typesupport when linking statically.

So having `rosidl_typesupport_introspection_c` and `rosidl_typesupport_introspection_cpp` available at the same time should not be a problem. But having more than one typesupport packages for the same language would be such a conflict.

**Comment by wjwwood** [comment]
> But having more than one typesupport packages for the same language would be such a conflict.

Thanks for clarify, that's important distinction. In that case I think this solution:

> - fail the build and ask the user to either switch to shared libraries or reduce the set of typesupports to only one

Is favorable.

**Comment by dirk-thomas** [comment]
@esteve Can you please comment here once comments have been addressed. Thanks!

**Comment by esteve** [comment]
@dirk-thomas @wjwwood I prefer to fail the build, instead of picking one typesupport implicitly or building shared libraries inconditionally. I've submitted https://github.com/ros2/rosidl_typesupport/pull/10 which aborts the build if there are multiple typesupports and the user wants to build ROS 2.0 statically (i.e. `BUILD_SHARED_LIBS` is disabled), let me know if that'd be fine. Thanks!

**Comment by dirk-thomas** [comment]
> I prefer to fail the build, instead of picking one typesupport implicitly or building shared libraries inconditionally.

Sounds good to me.

File: rosidl_generator_c/package.xml [hunk 3]
```diff
@@ -7,11 +7,11 @@
   <maintainer email="william@osrfoundation.org">William Woodall</maintainer>
   <license>Apache License 2.0</license>
 
-  <buildtool_depend>ament_cmake</buildtool_depend>
+  <buildtool_depend>ament_cmake_ros</buildtool_depend>
 
   <build_depend>rosidl_typesupport_interface</build_depend>
 
-  <buildtool_export_depend>ament_cmake</buildtool_export_depend>
+  <buildtool_export_depend>ament_cmake_ros</buildtool_export_depend>
```
Comment (dirk-thomas) [inline rosidl_generator_c/package.xml]: Why is this needed?

Same below.

Comment (esteve) [inline rosidl_generator_c/package.xml]: I incorrectly assumed that ament_cmake_ros had to be exported to other packages, but it doesn't seem to be the case. Removed in https://github.com/ros2/rosidl/pull/183/commits/f6bc5200d3ecf10129d8d81283822614516d3a67

File: rosidl_typesupport_introspection_c/cmake/rosidl_typesupport_introspection_c_generate_interfaces.cmake [hunk 4]
```diff
@@ -114,7 +114,7 @@ list(APPEND _generated_msg_header_files "${_visibility_control_file}")
 
 set(_target_suffix "__rosidl_typesupport_introspection_c")
 
-add_library(${rosidl_generate_interfaces_TARGET}${_target_suffix} SHARED
+add_library(${rosidl_generate_interfaces_TARGET}${_target_suffix}
```
Comment (dirk-thomas) [inline rosidl_typesupport_introspection_c/cmake/rosidl_typesupport_introspection_c_generate_interfaces.cmake]: Packages not using `ament_cmake_ros` will build this library static. Isn't that a problem?

File: rosidl_typesupport_introspection_cpp/cmake/rosidl_typesupport_introspection_cpp_generate_interfaces.cmake [hunk 5]
```diff
@@ -94,7 +94,7 @@ add_custom_command(
 
 set(_target_suffix "__rosidl_typesupport_introspection_cpp")
 
-add_library(${rosidl_generate_interfaces_TARGET}${_target_suffix} SHARED
+add_library(${rosidl_generate_interfaces_TARGET}${_target_suffix}
```
Comment (dirk-thomas) [inline rosidl_typesupport_introspection_cpp/cmake/rosidl_typesupport_introspection_cpp_generate_interfaces.cmake]: Same here.

File: rosidl_typesupport_introspection_c/cmake/rosidl_typesupport_introspection_c_generate_interfaces.cmake (hunk 4 above)
Comment (esteve) [inline rosidl_typesupport_introspection_c/cmake/rosidl_typesupport_introspection_c_generate_interfaces.cmake]: That's true, shall I update the messages packages so that they use `ament_cmake_ros` as well? Thanks for noticing.

**Comment by dirk-thomas** [comment]
Any news on this?

**Comment by esteve** [comment]
@dirk-thomas I was waiting on a response to https://github.com/ros2/rosidl/pull/183#discussion_r114950597 Do you think that updating the message packages to use `ament_cmake_ros` would be a valid solution to avoid the problem you described?

**Comment by dirk-thomas** [comment]
> Do you think that updating the message packages to use ament_cmake_ros would be a valid solution to avoid the problem you described?

I can't say for sure. I see the potential problem if parts of the system are being build with one setting and another part with the opposite. While that might not need to be a use case which works in all possible combinations the behavior should at least be reasonable for the user (e.g. fail with an error message if the combination can't work).

**Comment by mikaelarguedas** [comment]
@clalancette why is this tagged as untargeted?
I think we still want to do this for microcontrollers as well as embedded applications like android/iOS that require static linking

**Comment by esteve** [comment]
Closing in favor of #266 , thanks @dirk-thomas

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/announcing-objective-c-on-macos-and-ios-for-ros2/946/1
